rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,296 @@
+    /** If the result is a related result, provide the ID of the 'original' result */
+    related: string | null;
+    /** The media type being searched */
+    mediaType: MediaType;
+    /** The slug (not the prettified name) of the provider */
+    provider: string;
+    /** The search term **/
+    query: string;
+  };
+  /** The user clicks the CTA button to the external source to use the image */
+  GET_MEDIA: {
+    /** The unique ID of the media */
+    id: string;
+    /** The slug (not the prettified name) of the provider */
+    provider: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user clicks one of the buttons to copy the media attribution */
+  COPY_ATTRIBUTION: {
+    /** The unique ID of the media */
+    id: string;
+    /** The format of the copied attribution */
+    format: "plain" | "rich" | "html";
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user reports a piece of media through our form */
+  REPORT_MEDIA: {
+    /** The unique ID of the media */
+    id: string;
+    /** The slug (not the prettified name) of the provider */
+    provider: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user plays, pauses, or seeks an audio track.
+   * @todo: This potentially requires throttling.
+   **/
+  AUDIO_INTERACTION: {
+    /** The unique ID of the media */
+    id: string;
+    event: "seek" | "play" | "pause";
+    /** The slug (not the prettified name) of the provider */
+    provider: string;
+  };
+  /** The user visits a CC license description page on CC.org  **/
+  VISIT_LICENSE_PAGE: {
+    /** The slug of the license the user clicked on */
+    license: string;
+  };
+  /** The user visits a creator's link in the single result UI  **/
+  VISIT_CREATOR_LINK: {
+    /** The unique ID of the media */
+    id: string;
+    /** The permalink to the creator's profile */
+    url: string;
+  };
+  /** The user right clicks a single image result, most likely to download it. **/
+  RIGHT_CLICK_IMAGE: {
+    /** The unique ID of the media */
+    id: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The user uses the 'back to search' link on a single result **/
+  BACK_TO_SEARCH: {
+    /** The unique ID of the media */
+    id: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The visibility of the filter sidebar on desktop is toggled **/
+  TOGGLE_FILTERS: {
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+};
+type AnalyticsEventTypes = keyof AnalyticsEvents;
+
+/**
+ * Send an event to our analytics endpoint. Should be a no-op if the
+ * user has opted-out of analytics.
+ */
+async function sendEvent<T extends AnalyticsEventTypes>(
+  key: T,
+  payload?: AnalyticsEvents[T]
+): Promise<any> {
+  /**
+   * The default values to send with every analytics event.
+   */
+  const defaults = {
```

Comment:
I just realized, we should also be recording the language/locale in use. Also, if we retrieve some broad information about the IP such as country, but not actually store the IP, would it be considered invasive?

For reference, GoatCounter records both country and language.

Reply:
Just saw @obulat's comment https://github.com/WordPress/openverse/pull/312#discussion_r1041931113 above in a re-read.

Reply:
Added language tracking in 34fea2a57093c38c25ae218a8fc4ac197d3bea6d.